### ISSUE TITLE
Require taking self as argument for contract methods that interact with self

### ISSUE THREAD

**Issue by cburgdorf** [ethereum/fe#96]
## Proposed Change

Currently methods on contracts have implicit access to `self` as seen in the following snippet.

```
contract Something:
    x: u256 = 1
    pub def read_x() -> u256:
        return self.x
```

This issue proposes to explicitly require `self` as method argument when a method wants access to `self`:

```
contract Something:
    x: u256 = 1
    pub def read_x(self) -> u256:
        return self.x
```

## Motivation

One of the core motivations is to make `self` less *magical* (*where does it come from?*, *Why is it a special case?*). 
However there are other compelling reasons why such a change would be beneficial.

### Self-less methods as class methods

If a method does not interact with `self`, it can leave out `self` and becomes a *static* method of the contract.

```
contract Something:
    x: u256 = 1
    pub def does_not_read() -> u256:
        return 1 + 1
```

This is good for readability and auditing as it becomes easy to spot which methods do and do not interact with contract storage.
**Open Question**: Would such a method still be callable as `self.does_not_read()` or only as `Something.does_not_read()` ?

### Making mutability explicit

We can go one step further and say that `self` becomes immutable by default. If one wants to have a method that updates a storage variable, one has to explicitly *take `self`* with the `mut` keyword as shown below.

```
contract Something:
    x: u256 = 1
    pub def update_x(mut self) -> u256:
        self.x = 2
```

This again is great for readability and auditability. We can compare that to the inverse of `view` in solidity but I would argue that the nice aspect here is that *immutability* is the default and one has to *add* a keyword to make things mutable rather than to add a keyword to declare a method as being *view* only or *pure*.

**Comment by pipermerriam** [comment]
I **REALLY** like `mut self` and that entire mechanism for preventing state mutability.

I think it's worth exploring whether this can serve as re-entrancy protection as well.  If you went the rust route and were able to enforce there only ever being one mutable handle to `self` then it at least seems like that provides strong re-entrancy protection since a call frame holding the mutable handle to `self` would make any re-entrant calls fail as soon as they attempted to modify storage.

**Comment by cburgdorf** [comment]
>I think it's worth exploring whether this can serve as re-entrancy protection as well. If you went the rust route and were able to enforce there only ever being one mutable handle to self then it at least seems like that provides strong re-entrancy protection since a call frame holding the mutable handle to self would make any re-entrant calls fail as soon as they attempted to modify storage.

Yep, in `rust` a method that takes `mut self` can not even be called recursively (Because the second call would try get a mutable handle which isn't possible because the first call is still holding it). But if I understand correctly, this is a zero cost compile time guarantee that the compiler can make because it is in total control. In the Ethereum re-entrency case, other potentially unknown code can call into us which is something we can not prevent *free of cost* at compile time. We could still compile `mut self` down to something that enforces re-entrency protection, but it wouldn't be a zero cost abstraction. But maybe that is fine. E.g. the compiler can be smart enough to only add re-entrency protection to *public* methods that need it and leave it out for any other private methods that take `mut self` (it will still enforce the basic rules e.g. wouldn't be callable recursively).

I personally think that this is a very interesting idea.

**Comment by pipermerriam** [comment]
@g-r-a-n-t 

I have a question about this example:

```python
contract Foo:
  my_array: u256[100]

  def bar(self, mut mem_array: u256[100]):
    mem_array[26] = 42

  pub def baz(self):
    self.bar(self.my_array) # my_array is copied to memory
```

I would expect the compiler to throw an error on `self.bar(self.my_array)` because `self` should be immutable in that context, and thus, I would expect `self.my_array` to also be immutable, and since the function signature of `bar` specifies that it expects a mutable array type, that the type system would complain.  What does rust do in a case like this?

**Comment by g-r-a-n-t** [comment]
Ah, yeah good point. This is tricky.

In Rust, you would get the error you've described (assuming you don't have ownership issues).

We could enforce the same rules as Rust, but it is of course unnecessary since there is an implicit copy happening. I suppose we could consider adding some sort of copy method so it's more clear that you're not actually passing an immutable storage value.

```python
contract Foo:
  my_array: u256[100]

  def bar(self, mut mem_array: u256[100]):
    mem_array[26] = 42

  pub def baz(self):
    self.bar(self.my_array.to_mem()) # my_array is copied to memory
```

**Comment by g-r-a-n-t** [comment]
Also, I'm thinking in terms of ownership here, so `self.my_array.to_mem()` would be moved into `bar` where mutability is set by the parameter declaration.

This on the other hand would fail:

```python
pub def baz(self):
  mem_array: u256[100] = self.my_array.to_mem()
  self.bar(mem_array) # error: bar would mutate an immutable reference
  do_something(mem_array)
```

**Comment by cburgdorf** [comment]
>Are there instances where reentrance is desirable?

Good question. I have a hard time thinking of any but it's probably worth reaching out to people that regularly write smart contract code.

As a comparision, in Vyper one would [use the `@nonreentrant(<unique_key>)` decorator](https://vyper.readthedocs.io/en/v0.1.0-beta.17/structure-of-a-contract.html#decorators) to protect against reentrancy.

But, I like the fact that we wouldn't need something like a special decorator for this because it would keep the language simpler and safer after all.

 >If so, how should we allow it in the language (assuming it's disabled by default)?

One idea would be to introduce something like [`RefCell`](https://doc.rust-lang.org/std/cell/struct.RefCell.html#method.borrow_mut) which could give more fine grained mutability via something like `borrow_mut`. But honestly, that's a route of complexity that I'd rather like to avoid.

Another idea would be something like a decorator `@allow_reentrancy` but I'm still skeptical about any valid use cases. I would like to see a case that would really need a state mutating function to be reentrant and couldn't be rewritten another way.  

>For now, I think the simplest thing we can do is require that all parameters be passed by value or memory pointer

Yeah, I think I'm on board :+1:

**Comment by cburgdorf** [comment]
Here's a case from Aragon: https://mobile.twitter.com/izqui9/status/1317826431527886849

**Comment by cburgdorf** [comment]
Ignoring all the mutability stuff for now but I think we should prioritize making `self` mandatory for non-static functions sooner than later. The reason for that is that it may proof very useful related to `struct` support which is coming with #203.

One of the planned features for structs is to allow functions on structs (just like in Rust). Now, functions on structs could be very useful for both as instance methods or static functions as demonstrated with the following example.

```
struct Rectangle:
    width: u256
    height: u256

# Static method to create a rectangle that happens to be a square    
def square(size: u256) -> Rectangle:
    Rectangle(width=size, height: size)

# An instance method to check if the rectangle can hold another rectangle
def can_hold(self, other: Rectangle) -> bool:
    return self.width > other.width and self.height > other.height
```

Taking `self` as the first parameter is what would decide whether something is a static method or an instance method (just like in Rust)

Usage:

```
big = Rectangle(width=10, height=20)
square = Rectangle::square(5)
```
